### ISSUE TITLE
Load generator CO free and HdrHistogram

### ISSUE THREAD

**Issue by franz1981** [HdrHistogram/HdrHistogram#148]
HI Gil,

sorry to ask this another time, but I'm not able to fully understand the relation of HdrHistogram CO compensation and a proper (hopefully) CO free load generator.
If I'm already using a load generator that would trigger requests according to a defined schedule (ie that won't "coordinate" with the measured system, trying to stick to a fixed rate, but speeding it up to keepup if necessary) and I'm already using HdrHistogram::recordValue using the response time ie (endOfOperation - intendedStart):

1. am I using HdrHistogram in the right way? 
2. am I correctly measuring responsiveness under load with it? 

Thanks for every explanation you give :)

NOTE:
I will put this issue/question link on mechanical sympathy to get anyone interested in, able to comment or just read it :)

**Comment by giltene** [comment]
For load generated that works as you described, especially the recordValue(endOfOperation - intendedStart) part (as opposed to the more typical recordValue(endOfOperation - actualStart)), you are doing it right, there is no need to use [recordValueWithExpectedInterval](http://hdrhistogram.github.io/HdrHistogram/JavaDoc/index.html?org/HdrHistogram/package-summary.html)(). 

I claim that all load generators that "have a plan" can be made to correctly record latencies this way, but monitoring is quite different: there is no "plan" from which one can extract the intendedStart value. Luckily, it is often possible to "reconstruct" known-to-be-missing events that should have occurred at a certain pace or expected interval, but were not observed and cannot be re-observed. This reconstruction is quite often done based on an expected interval modeling, and that is where 
recordValueWithExpectedInterval() is quite useful.

For a simple and very practical example of where recordValueWithExpectedInterval is useful this way, look at how [jHiccup](https://github.com/giltene/jHiccup/blob/master/src/main/java/org/jhiccup/HiccupMeter.java#L488) uses it. jHiccup monitors actual observed hiccups, but hiccups longer than the intended interval cause skips in observations. Luckily, since the expected interval between observations is known, and all hiccups that span multiple intervals will always result in lost samples, the magnitude of the lost sample can be conservatively estimated (i.e. we know of "missing hiccups of at least this size"), and recordValueWithExpectedInterval() can be used directly and the caller's code is kept simple.

For a more elaborate example, see LatencyUtils's use in a pause detector [here](https://github.com/LatencyUtils/LatencyUtils/blob/d8f51f39f6146e1ad9a263dc916bcbc0ec06e16d/src/main/java/org/LatencyUtils/LatencyStats.java#L405). LatencyUtils employs a pluggable pause detector (and supplies one that is pretty good at reliably identifying process-wide pauses, and separating them from thread-local glitches) along with an interval estimator, and recreates presumed-missing latency samples whenever a pause is detected. It is useful in "fairly reliably and conservatively" recovering missing latency measurements due to internal coordinated omission effects in self monitoring processes that are caused by process pauses (pretty much anywhere where you measure the time before and after an operation to record it's latency, but are exposed to process-wide pauses), but it is not able to compensate for coordinated omission that is caused by "falling behind" and building up queues (in TCP, in load balancers, in client processes)...

**Comment by franz1981** [comment]
Thanks @marshallpierce for the super fast response: so probably I'm a similar but different situation...
Consider that you have a `pushLoad` call that can block for an undefined amount of time and you want to measure the responsiveness of such pushLoad call, if you have a single threaded loop like this:
```
long nextIntendedStart = ...
while(true) {
    long next = waitUntilNextScheduledFireTime(nextIntendedStart);
    pushLoad();
    long now = System.nanoTime;
    histogram.recordValue(now - nextIntendedStart);
    nextIntendedStart = next;
}
```
Such a load generator will try (as any load generators can do) to keepup with the expected schedule but obviously it won't push effectively the load when the measured system is busy.
With this type of load generator (assuming that's correct!), it is fine to measured response time in that way with HdrHistogram?

**Comment by marshallpierce** [comment]
IIUC, in that usage, you would need to correct for CO, since if `pushLoad()` took an abnormally long time to execute (e.g. full GC), you would have missed potentially many slow calls that, if they were accurately accounted for, would (correctly) skew your response time distribution to have a much higher 99th pctile, etc.

**Comment by giltene** [comment]
@franz1981 the key in the code sample above is that the value that waitUntilNextScheduledFireTime(nextIntendedStart); returns needs to be the next intended (according to the model of what the load is supposed tube doing) time, and not the time of the return. Specifically, that time should often be "in the past". As long as that is true, I think your loop is fine.

Actually providing this computation is not always trivial. If all we have is single request/response stuff with no dependencies between requests, the math to determine the next intended time is very simple. However, when the request you send is dependent on the contents of a previous response (which is quite often the case in actual application logic), it's intended start time depends on both the intended start time of the request it depends on, and the "expected response time" of that request. If you use the actual response time for the previous request (that this request depend on), you will still have CO. Computing the "Expected response time" often requires some estimation code (e.g. measure the median response time for that sort of request and use that as the "expected response time") since, people usually don't supply this as part of the "plan".

**Comment by giltene** [comment]
> IIUC, in that usage, you would need to correct for CO, since if `pushLoad()` took an abnormally long time to execute (e.g. full GC)...

if waitUntilNextScheduledFireTime(nextIntendedStart) correctly computes the returned next value, and allows it to be in the past, and obviously not wait when given a time in the past, the logic should work fine.

**Comment by franz1981** [comment]
@giltene @marshallpierce 

> if waitUntilNextScheduledFireTime(nextIntendedStart) correctly computes the returned next value, and allows it to be in the past, and obviously not wait when given a time in the past, the logic should work fine.

that's exactly what it is supposed to do :+1:

**Comment by marshallpierce** [comment]
> if waitUntilNextScheduledFireTime(nextIntendedStart) correctly computes the returned next value, and allows it to be in the past, and obviously not wait when given a time in the past, the logic should work fine.

Good point. Does this assume that the system's behavior is roughly equivalent when requests that should have started during a pause are simply started late rather than starting on time and being queued? Or, put another way, while correcting for CO is clearly better than nothing, do you have a sense of how close it is to the measurements you would obtain by generating load -- presumably with nonblocking i/o -- where requests really are started on schedule always, regardless of extant requests? I haven't actually tried writing a nonblocking generator *and* a blocking one with CO correction to see.

**Comment by giltene** [comment]
> > if waitUntilNextScheduledFireTime(nextIntendedStart) correctly computes the returned next value, and allows it to be in the past, and obviously not wait when given a time in the past, the logic should work fine.
> 
> Good point. Does this assume that the system's behavior is roughly equivalent when requests that should have started during a pause are simply started late rather than starting on time and being queued? Or, put another way, while correcting for CO is clearly better than nothing, do you have a sense of how close it is to the measurements you would obtain by generating load -- presumably with nonblocking i/o -- where requests really are started on schedule always, regardless of extant requests? I haven't actually tried writing a nonblocking generator _and_ a blocking one with CO correction to see.

In any case, you won't have coordinated **omission** if you do all your measurements this way, as all planned operations are actually measured, and their times are measured from the planned start time to the actual operation completion time (so no omission). 

You could have a difference in modeling the load tho, as the "catch up" operations are started later than planned, which may make a difference to the system's behavior under load. To make sure you load happens "exactly as planned", you would need to actually enqueue the operations to the tested system at the planned times. However, this (getting requests to hit tested shrive on planned times) is usually impractical for all but the trivial setups, as request to request dependency almost always exists somewhere (e.g. in TCP receive and send windows that will eventually backlog into your load generator sockets).

Bottom line: you can make sure that the response time measurements reflect what actually happened under the load that you made happen. This is not always what would actually happen if the planned load actually happened, but it is much better than if you let CO do its thing.

**Comment by franz1981** [comment]
@giltene @marshallpierce So we can just patch the omission part, but some kind of *coordination* with any moving parts that will end with the load generation is simply unavoidable....

**Comment by franz1981** [comment]
Thanks guys about it: now I've more clear when use reacordValue with expected interval and when not :+1:
